### PR TITLE
Dynamic batch-size sampler based on `num_nodes`/`num_edges`

### PR DESCRIPTION
**Pull request by GericoVi**
As discussed in #583

### CODE REVIEW

File: torch_geometric/loader/__init__.py [hunk 1]
```diff
@@ -12,23 +12,13 @@
...
-    'DenseDataLoader',
-    'TemporalDataLoader',
-    'NeighborSampler',
-    'ImbalancedSampler',
+    'DataLoader', 'NeighborLoader', 'LinkNeighborLoader', 'HGTLoader',
+    'ClusterData', 'ClusterLoader', 'GraphSAINTSampler',
+    'GraphSAINTNodeSampler', 'GraphSAINTEdgeSampler',
+    'GraphSAINTRandomWalkSampler', 'ShaDowKHopSampler', 'RandomNodeSampler',
+    'DataListLoader', 'DenseDataLoader', 'TemporalDataLoader',
+    'NeighborSampler', 'ImbalancedSampler', 'DynamicBatchSampler'
```
Comment (rusty1s) [inline torch_geometric/loader/__init__.py:23]:
```suggestion
    'NeighborSampler', 'ImbalancedSampler', 'DynamicBatchSampler',
```
to let `yapf` re-format it.

File: torch_geometric/loader/dynamic_batch_sampler.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
+from typing import Iterator, List, Optional
+
+import torch
+
+from torch_geometric.data import Dataset
+
+
+class DynamicBatchSampler(torch.utils.data.sampler.Sampler[List[int]]):
+    r"""Dynamically adds samples to mini-batch up to a maximum size (number of
```
Comment (rusty1s) [inline torch_geometric/loader/dynamic_batch_sampler.py:9]:
```suggestion
    r"""Dynamically adds samples to a mini-batch up to a maximum size (number of
```

File: torch_geometric/loader/dynamic_batch_sampler.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+    ``num_steps`` can be supplied to cap the number of mini-batches produced
+    by the sampler.
+
+    **Usage:**
+
+    .. code-block:: python
+
+        from torch_geometric.loader import DataLoader, DynamicBatchSampler
+
+        batch_sampler = DynamicBatchSampler(10000, dataset)
```
Comment (rusty1s) [inline torch_geometric/loader/dynamic_batch_sampler.py:26]:
```suggestion
        batch_sampler = DynamicBatchSampler(dataset, num_nodes=10000)
```
Can we first pass in the data, and also rename `max_num` to `num_nodes`?

Comment (GericoVi) [reply torch_geometric/loader/dynamic_batch_sampler.py:26]: Yes, the data argument first makes sense. For the other, I had it as `max_num` instead of `num_nodes` since you can also specify a maximum number of edges in the batch instead of nodes with the `entity` argument. Is there a more generic name? Or do you think `num_nodes` is fine?

File: test/loader/test_dynamic_batch_sampler.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+import torch
+
+from torch_geometric.data import Data
+from torch_geometric.loader import DataLoader, DynamicBatchSampler
+
+
+def test_dataloader_with_dynamic_batches():
+    data_list: List[Data] = []
+    for i in range(100, 110):
```
Comment (rusty1s) [inline test/loader/test_dynamic_batch_sampler.py:11]:
```suggestion
    for num_nodes in range(100, 110):
```

File: test/loader/test_dynamic_batch_sampler.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        data_list.append(Data(num_nodes=i))
+
+    torch.manual_seed(12345)
+    batch_sampler = DynamicBatchSampler(data_list, 300, shuffle=True)
+    loader = DataLoader(data_list, batch_sampler=batch_sampler)
+
+    num_nodes_total = 0
+    for data in loader:
+        num_nodes_total += data.num_nodes
```
Comment (rusty1s) [inline test/loader/test_dynamic_batch_sampler.py:20]: Add an assert that the number of nodes per batch is not larger than 300.

File: torch_geometric/loader/dynamic_batch_sampler.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
+from typing import Iterator, List, Optional
+
+import torch
+
+from torch_geometric.data import Dataset
+
+
+class DynamicBatchSampler(torch.utils.data.sampler.Sampler[List[int]]):
+    r"""Dynamically adds samples to a mini-batch up to a maximum size (number of
+    nodes). When data samples have a wide range in sizes, specifying a
```
Comment (rusty1s) [inline torch_geometric/loader/dynamic_batch_sampler.py:10]: Either based on number of nodes or number of edges.

File: torch_geometric/loader/dynamic_batch_sampler.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    mini-batch size in terms of number of samples is not ideal and can cause
+    CUDA OOM errors.
+
+    Therefore, the number of steps per epoch is ambiguous, depending on the
```
Comment (rusty1s) [inline torch_geometric/loader/dynamic_batch_sampler.py:14]:
```suggestion
    Within the :class:`DynamicBatchSampler`, the number of steps per epoch is ambiguous, depending on the
```

File: torch_geometric/loader/dynamic_batch_sampler.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    order of the samples. By default the ``__len__()`` will be None, fine for
```
Comment (rusty1s) [inline torch_geometric/loader/dynamic_batch_sampler.py:15]:
```suggestion
    order of the samples. By default the :meth:`__len__` will be undefined. This is fine for
```

File: torch_geometric/loader/dynamic_batch_sampler.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    most cases but progress bars will be `infinite`. Alternatively,
+    ``num_steps`` can be supplied to cap the number of mini-batches produced
```
Comment (rusty1s) [inline torch_geometric/loader/dynamic_batch_sampler.py:17]:
```suggestion
    :obj:`num_steps` can be supplied to cap the number of mini-batches produced
```

File: torch_geometric/loader/dynamic_batch_sampler.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+    .. code-block:: python
+
+        from torch_geometric.loader import DataLoader, DynamicBatchSampler
+
+        batch_sampler = DynamicBatchSampler(dataset, max_num=10000)
+        loader = DataLoader(dataset, batch_sampler=sampler, ...)
+
+    Args:
+        data_source (Dataset): dataset to sample from, need this so that
```
Comment (rusty1s) [inline torch_geometric/loader/dynamic_batch_sampler.py:30]:
```suggestion
        dataset (Dataset): Dataset to sample from.
```

File: torch_geometric/loader/dynamic_batch_sampler.py [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+            potential samples can be sized to create the dynamic batch.
+        max_num (int): size of mini-batch to aim for in number of nodes or
+            edges.
+        mode (str, optional): `nodes` or `edges` to measure batch size.
```
Comment (rusty1s) [inline torch_geometric/loader/dynamic_batch_sampler.py:34]:
```suggestion
        mode (str, optional): obj:`node` or obj:`edge` to measure batch size.
```

File: torch_geometric/loader/dynamic_batch_sampler.py [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+            (default: `nodes`)
```
Comment (rusty1s) [inline torch_geometric/loader/dynamic_batch_sampler.py:35]:
```suggestion
            (default: obj:`node`)
```

File: torch_geometric/loader/dynamic_batch_sampler.py [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+        shuffle (bool, optional): set to ``True`` to have the data reshuffled
```
Comment (rusty1s) [inline torch_geometric/loader/dynamic_batch_sampler.py:36]: See other occurrences and replace with `:obj:` to let the doc render uniformly.

File: torch_geometric/loader/dynamic_batch_sampler.py [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+            at every epoch (default: ``False``).
+        skip_too_big (bool, optional): set to ``True`` to skip samples which
+            can't fit in a batch by itself. (default: ``False``).
+        num_steps (int, optional): The number of mini-batches to draw for a
+            single epoch. If set to :obj:`None`, will iterate through all the
+            underlying data, but __len__ will be :obj:`None` since it will be
```
Comment (rusty1s) [inline torch_geometric/loader/dynamic_batch_sampler.py:42]:
```suggestion
            underlying data, but :meth:`__len__` will be :obj:`None` since it will be
```

File: torch_geometric/loader/dynamic_batch_sampler.py [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+            ambiguous. (default: :obj:`None`)
+    """
+    def __init__(self, data_source: Dataset, max_num: int, mode: str = 'nodes',
+                 shuffle: bool = False, skip_too_big: bool = False,
+                 num_steps: Optional[int] = None):
+        if not isinstance(max_num, int) or max_num <= 0:
+            raise ValueError("`max_num` should be a positive integer value, "
+                             "but got max_num={max_num}.")
```
Comment (rusty1s) [inline torch_geometric/loader/dynamic_batch_sampler.py:50]:
```suggestion
                             "(got {max_num)}.")
```

File: torch_geometric/loader/dynamic_batch_sampler.py [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+            num_steps = len(data_source)
+        self.num_steps = num_steps
+
+        self.mode = f"num_{mode}"
+
+    def __iter__(self) -> Iterator[List[int]]:
+        batch = []
+        batch_n = 0
+        num_steps = 0
+        current_place = 0
```
Comment (rusty1s) [inline torch_geometric/loader/dynamic_batch_sampler.py:69]:
```suggestion
        num_processed = 0
```
I find `current_place` a bit misleading due to shuffle option.

File: torch_geometric/loader/dynamic_batch_sampler.py [hunk 17]
```diff
@@ -0,0 +1,108 @@
...
+        else:
+            self.idxs = torch.arange(len(self.data_source), dtype=torch.long)
+
+        # Main iteration loop
+        while (current_place < len(self.data_source)
+               and num_steps < self.num_steps):
+            # Fill batch
+            for idx in self.idxs[current_place:]:
+                # Size of sample
+                n = getattr(self.data_source[idx], self.mode)
```
Comment (rusty1s) [inline torch_geometric/loader/dynamic_batch_sampler.py:82]: Can we make this explicit here and branch out based on mode?

File: torch_geometric/loader/dynamic_batch_sampler.py [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+
+                if batch_n + n > self.max_num:
+                    if batch_n == 0:
+                        if self.skip_too_big:
+                            continue
+                        else:
+                            raise RuntimeError(
```
Comment (rusty1s) [inline torch_geometric/loader/dynamic_batch_sampler.py:89]: Won‘t we want to skip instead?

File: torch_geometric/loader/dynamic_batch_sampler.py [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+        self.mode = f"num_{mode}"
+
+    def __iter__(self) -> Iterator[List[int]]:
+        batch = []
+        batch_n = 0
+        num_steps = 0
+        current_place = 0
+
+        if self.shuffle:
+            self.idxs = torch.randperm(len(self.data_source), dtype=torch.long)
```
Comment (rusty1s) [inline torch_geometric/loader/dynamic_batch_sampler.py:72]:
```suggestion
            indices = torch.randperm(len(self.data_source), dtype=torch.long)
```
Do we need to add it to `self`?

**Review by rusty1s — COMMENTED** [review]
Left some final (nitpicky) comments. Thank you!

File: torch_geometric/loader/dynamic_batch_sampler.py (hunk 18 above)
Comment (GericoVi) [reply torch_geometric/loader/dynamic_batch_sampler.py:91]: I don't quite understand this comment. Essentially the purpose of that bit is that if the batch is empty and the next sample will overfill it - either skip that sample or raise an error.
If it's not empty, currently it outputs the batch and then attempts to use that sample in the next batch.

Comment (rusty1s) [reply torch_geometric/loader/dynamic_batch_sampler.py:91]: That makes sense! Thanks!

Comment (rusty1s) [reply torch_geometric/loader/dynamic_batch_sampler.py:91]: I looked at it again and feel that we should output a warning here and still add that data object in (as a single example per batch). Currently, we only have the option to skip or to error out - and neither of this may be what the user desires.

**Review by rusty1s — APPROVED** [review]
Thanks for the updates. Found some minor last issues but happy to merge afterwards. Left one last comment regarding erroring out on too big examples - I think we should print out a warning but still keep the example. WDYT?

**Comment by rusty1s** [comment]
Can you also add this PR to the changelog?

File: torch_geometric/loader/dynamic_batch_sampler.py [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+    wide range in sizes, specifying a mini-batch size in terms of number of
+    samples is not ideal and can cause CUDA OOM errors.
+
+    Within the :class:`DynamicBatchSampler`, the number of steps per epoch is
+    ambiguous, depending on the order of the samples. By default the
+    :meth:`__len__` will be undefined. This is fine for most cases but
+    progress bars will be `infinite`. Alternatively, :obj:`num_steps` can be
+    supplied to cap the number of mini-batches produced by the sampler.
+
+    **Usage:**
```
[the remark this answers is not in the document]
Comment (GericoVi) [reply torch_geometric/loader/dynamic_batch_sampler.py:20]: Which header, sorry? The usage portion?

File: torch_geometric/loader/dynamic_batch_sampler.py (hunk 18 above)
Comment (GericoVi) [reply torch_geometric/loader/dynamic_batch_sampler.py:91]: Yes that probably sounds like a better idea actually. So instead of error, just throw a warning and add the example anyway? To clarify, when this happens there would only be a single example in the mini-batch, and `data.num_nodes` will be greater than the `max_num` specified for that one mini-batch.

That sounds reasonable to me.

Comment (rusty1s) [reply torch_geometric/loader/dynamic_batch_sampler.py:91]: Super :)

File: torch_geometric/loader/dynamic_batch_sampler.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (rusty1s) [reply torch_geometric/loader/dynamic_batch_sampler.py:20]: Yes, indeed.

[the remark this answers is not in the document]
Comment (GericoVi) [reply torch_geometric/loader/dynamic_batch_sampler.py:20]: Makes sense, will do.

**Comment by GericoVi** [comment]
Discussed updates done. Feel free to merge if happy with changes.

**Review by rusty1s — APPROVED** [review]
